lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUTransformCharFilter.java

```diff
@@ -0,0 +1,322 @@
+
+/**
+ * A {@link CharFilter} that transforms text with ICU.
+ * <p>
+ * ICU provides text-transformation functionality via its Transliteration API.
+ * Although script conversion is its most common use, a Transliterator can
+ * actually perform a more general class of tasks. In fact, Transliterator
+ * defines a very general API which specifies only that a segment of the input
+ * text is replaced by new text. The particulars of this conversion are
+ * determined entirely by subclasses of Transliterator.
+ * </p>
+ * <p>
+ * Some useful transformations for search are built-in:
+ * <ul>
+ * <li>Conversion from Traditional to Simplified Chinese characters
+ * <li>Conversion from Hiragana to Katakana
+ * <li>Conversion from Fullwidth to Halfwidth forms.
+ * <li>Script conversions, for example Serbian Cyrillic to Latin
+ * </ul>
+ * <p>
+ * Example usage: <blockquote>stream = new ICUTransformCharFilter(reader,
+ * Transliterator.getInstance("Traditional-Simplified"));</blockquote>
+ * <br>
+ * For more details, see the <a
+ * href="http://userguide.icu-project.org/transforms/general">ICU User
+ * Guide</a>.
+ */
+public final class ICUTransformCharFilter extends BaseCharFilter {
+
+  // Transliterator to transform the text
+  private final Transliterator transform;
+
+  // Reusable position object
+  private final Position position = new Position();
+
+  private static final int READ_BUFFER_SIZE = 1024;
+  private final char[] tmpBuffer = new char[READ_BUFFER_SIZE];
+
+  private static final int INITIAL_TRANSLITERATE_BUFFER_SIZE = 1024;
+  private final StringBuffer buffer = new StringBuffer(INITIAL_TRANSLITERATE_BUFFER_SIZE);
+  private final ReplaceableString replaceable = new ReplaceableString(buffer);
+
+  private static final int BUFFER_PRUNE_THRESHOLD = 1024;
+
+  private int outputCursor = 0;
+  private boolean inputFinished = false;
+  private int charCount = 0;
+
+  static final int DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY = 8192;
+  private final int maxRollbackBufferCapacity;
+
+  private static final int DEFAULT_INITIAL_ROLLBACK_BUFFER_CAPACITY = 4; // must be power of 2
+  private char[] rollbackBuffer;
+  private int rollbackBufferSize = 0;
+
+  ICUTransformCharFilter(Reader in, Transliterator transform) {
+    this(in, transform, DEFAULT_MAX_ROLLBACK_BUFFER_CAPACITY);
+  }
+
+  /**
+   * Construct new {@link ICUTransformCharFilter} with the specified {@link Transliterator}, backed by
+   * the specified {@link Reader}.
+   * @param in input source
+   * @param transform used to perform transliteration
+   * @param maxRollbackBufferCapacityHint used to control the maximum size to which this
+   * {@link ICUTransformCharFilter} will buffer and rollback partial transliteration of input sequences.
+   * The provided hint will be converted to an enforced limit of "the greatest power of 2 (excluding '1')
+   * less than or equal to the specified value". Specifying a negative value allows the rollback buffer to
```

Comment:
Is there any reason to support negative values, rather than having the caller supply MAX_VALUE? I think elsewhere in this codebase we tend to error on negative values in cases like this.

Reply:
+1. This was here more as a convenience targeted at the external schema config API, so I've preserved that convenience, but moved it out to the `ICUTransformCharFilterFactory` instead (to allow users who essentially want to configure "no limit" to avoid having to explicitly write something like "2147483647" in their config files). Does that sound ok?

I also noticed that there is no power of 2 greater than or equal to hint, for hint greater than `Integer.highestOneBit(Integer.MAX_VALUE)`, so I added special handling for that range of hint. It's a little fiddly, but basically we're protecting against overflow while recognizing that if the user specifies a very large value, it makes sense to resolve to the largest _legal_ value, even though it's smaller than the specified hint.